APF-2667 Update Discovery Response

jonathanrfisher1 opened the pull request:
Updated action label to specify local
Added mock config data

Commit:
APF-2667 Update Discovery Response

Updated action label to specify local
Added mock config data

RyanBard commented on samples/node/routes/discovery.js:
```diff
@@ -17,19 +17,35 @@ exports.root = function (req, res) {
                 endpoint: {href: `${base}/cards/requests`}
             }
+        },
+        // Logic for this config not implemented.  Value is present to provide mock config data in discovery.
```
It would only take a minor tweak to make this functional:

Change:
```
const zips = req.body.tokens.zip || [];
```

to:

```
const zips = req.body.tokens.zip || (req.body.config.defaultZip ? [req.body.config.defaultZip] : []);
```

jonathanrfisher1 replied:
Yeah, I’ll do it.  Would be nice to have it working.

RyanBard replied:
You've implemented it, but left the comment in saying it's not implemented.

Commit:
APF-2667 Discovery Changes

Consume zip from config.
Use config if no zip provided in tokens

RyanBard commented on samples/node/routes/weather.js:
```diff
@@ -9,19 +9,26 @@ const uuidV4 = require('uuid/v4');
 exports.requestCards = function(req, res) {
-    if (!req.body.tokens) {
-        res.status(400).send("Missing tokens field");
+    let tokenZip = (req.body.tokens && req.body.tokens.zip) ? req.body.tokens.zip : null;
```
There's no reason to use a ternary here.

```
let tokenZip = req.body.tokens && req.body.tokens.zip;
```

tokenZip will be falsy if tokens or tokens.zip is falsy.

Ditto for configZip.

jonathanrfisher1 replied:
I'm actually grabbing the array here, not just resolving to a boolean

RyanBard replied:
So tokenZip would be the array, or falsy.  My point still stands.

If you are concerned that they pass in an empty zip array, then you already aren't handling that properly (`!tokenZip && !configZip` would have to also check `.length`).

RyanBard replied:
Forget the ditto for configZip, however.  That still needs to be ternary (or something different), since it's converting the string to an array of the string.

jonathanrfisher1 replied:
Alright, check the new changes.  Updated the JS and fleshed out config+validation more.

RyanBard commented on samples/node/routes/weather.js:
```diff
@@ -9,19 +9,26 @@ const uuidV4 = require('uuid/v4');
+    console.log(`tokenZip: ${tokenZip}`);
+    console.log(`configZip: ${configZip}`);
+
+    const zips = tokenZip ? tokenZip : configZip;
```
There's no reason for a ternary here.  Just do this:

```
const zips = tokenZip || configZip;
```

EDIT: typo

jonathanrfisher1 replied:
These are not booleans, they are the actual objects. This doesn't work for actual objects right?

RyanBard replied:
Javascript is a dynamically + weakly typed language.  It doesn't care if it's a boolean or not, only that it's truthy/falsy.

Also, you said above that it's the arrays (it is), but now you're saying objects.  Usually you don't call arrays objects in javascript (they are, however it's usually -- but not always -- important to distinguish the difference).

RyanBard reviewed (changes requested):
* Simplify the JS
* Remove the comment saying you didn't implement config (b/c you did)

RyanBard commented:
Also, you have conflicts.  I think you forgot to fetch + rebase before you started (the conflicted file was last changed in June).

jonathanrfisher1 commented:
Hmm, I’ll check on that and apply my changes to a more recent version.

I thought I pulled and branched for this.

Commit:
APF-2667 Discovery Changes

Remove unnecessary comment

jonathanrfisher1 committed:
Merge branch 'master' into topic/fisher/APF-2667-Discovery

RyanBard reviewed (changes requested):
Open your browser dev tools and try some of these out:

```
console.log(false && 'foo');
console.log(true && 'bar');
console.log(false || 'baz');
console.log(true || 'abc');
console.log('hello' || 'world');
console.log('abc' && 'def');
```

* https://developer.mozilla.org/en-US/docs/Glossary/Truthy
* https://developer.mozilla.org/en-US/docs/Glossary/Falsy
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_Operators

Commit:
APF-2667 Discovery Update

Update config to be a fully fleshed example with validation
Add input validation for requests matching validators
Remove old 'foo' config example
Clean up JS

RyanBard commented on samples/node/routes/weather.js:
```diff
@@ -11,19 +11,37 @@ const sha1 = require('sha1');
+    let tokenZip = req.body.tokens && req.body.tokens.zip;
+    let configZip = req.body.config && req.body.config.defaultZip;
+
+    console.log(`tokenZip: ${tokenZip}`);
```
Do you need to log these if the whole request body was already logged (line 12)?

jonathanrfisher1 replied:
I don’t have to, I thought about cutting some of the logging out.

I was logging just to clearly show what values were parsed out for each.

RyanBard replied:
```
cards called:  { config: { defaultZip: '30332' },
  zips: [ '30080', '30076' ],
  connector_id: '46a42fb8-1097-427a-a8b1-491c8b53aba7',
  object_type: 'card' }
tokenZip: 30080,30076
configZip: 30332
Valid Input: 30080,30076
Zips: 30080,30076
```

and

```
cards called:  { config: { defaultZip: '30332' },
  connector_id: '46a42fb8-1097-427a-a8b1-491c8b53aba7',
  object_type: 'card' }
tokenZip: undefined
configZip: 30332
Valid Input: 30332
Zips: 30332
```

seem a little excessive, but it's not worth waiting another day or 2.
